runtime/src/main/java/org/corfudb/util/Utils.java

```diff
@@ -307,34 +319,29 @@ static TailsResponse aggregateLogUnitTails(Set<TailsResponse> responses) {
      * @return response with all streams addresses and global log tail.
      */
     public static StreamsAddressResponse getLogAddressSpace(RuntimeLayout runtimeLayout) {
-        Set<StreamsAddressResponse> luResponses = new HashSet<>();
-
-        Layout.LayoutSegment segment = runtimeLayout.getLayout().getLatestSegment();
-
-        // Query the head log unit in every stripe.
-        if (segment.getReplicationMode() == Layout.ReplicationMode.CHAIN_REPLICATION) {
-            for (Layout.LayoutStripe stripe : segment.getStripes()) {
-
-                StreamsAddressResponse res = CFUtils.getUninterruptibly(runtimeLayout
-                                .getLogUnitClient(stripe.getLogServers().get(DEFAULT_LOGUNIT))
-                                .getLogAddressSpace());
-                luResponses.add(res);
-            }
-        } else if (segment.getReplicationMode() == Layout.ReplicationMode.QUORUM_REPLICATION) {
-            throw new UnsupportedOperationException();
-        }
-
-        return aggregateLogAddressSpace(luResponses);
-    }
-
-    static StreamsAddressResponse aggregateLogAddressSpace(Set<StreamsAddressResponse> responses) {
-        Map<UUID, StreamAddressSpace> streamAddressSpace = new HashMap<>();
-        long logTail = Address.NON_ADDRESS;
+        // Since a node can exist as a head for multiple segments we need to a set to
+        // coalesce the candidates to unique nodes only
+        Set<String> segmentsHeadNodes = getChainHeadFromAllSegments(runtimeLayout.getLayout());
+        AtomicLong globalTail = new AtomicLong(Address.NON_EXIST);
+        final Map<UUID, StreamAddressSpace> streamsAddressSpace = new HashMap<>();
+        List<CompletableFuture<StreamsAddressResponse>> cfs = segmentsHeadNodes.stream()
+                .map(node -> runtimeLayout
+                        .getLogUnitClient(node)
+                        .getLogAddressSpace())
+                .collect(Collectors.toList());
 
-        for (StreamsAddressResponse res : responses) {
-            logTail = Math.max(logTail, res.getLogTail());
-            streamAddressSpace = aggregateStreamAddressMap(res.getAddressMap(), streamAddressSpace);
-        }
-        return new StreamsAddressResponse(logTail, streamAddressSpace);
+        cfs.stream().map(CFUtils::getUninterruptibly)
+                .forEach(resp -> {
+                    // All responses should be computed on the same epoch
+                    checkArgument(resp.getEpoch() == runtimeLayout.getLayout().getEpoch());
```

Comment:
This will throw IllegalArgumentException, but should be throw WrongEpochException to be more clear?

Reply:
good catch.